intelmq/bots/parsers/twitter/parser.py

```diff
@@ -0,0 +1,90 @@
+There is also a whitelist for filtering out good domains.
+
+
+Parameters:
+    domain_whitelist : domains that will be ignored in parsing
+    substitutions : semicolon separated list of pairs
+                    substitutions that will be made in the text,
+                    for example " .com,.com" enables parsing of one fuzzy format
+                                "[.];." enables the parsing of another fuzzy format
+
+"""
+
+from url_normalize import url_normalize
+from tld import get_tld
+from tld.utils import update_tld_names
+from intelmq.lib.bot import ParserBot
+from intelmq.lib.bot import utils
+import re
+
+
+class TwitterParserBot(ParserBot):
+    def init(self):
+        update_tld_names()
+        self.domain_whitelist = []
+        if getattr(self.parameters, "domain_whitelist", '') is not '':
+            self.domain_whitelist.extend(self.parameters.domain_whitelist.split(','))
+        self.substitutions = []
+        if getattr(self.parameters, "substitutions", '') is not '':
+            temp = self.parameters.substitutions.split(';')
+            if len(temp) % 2 != 0:
+                raise exceptions.InvalidArgument(
+                        'substitutions',
+                        got=self.parameters.substitutions,
+                        expected="even number of strings")
+            for i in range(int(len(temp)/2)):
+                self.substitutions.append([temp[2*i], temp[2*i+1]])
+
+    def get_domain(self, address):
+        try:
+            dom = re.search('(/|^)([a-z0-9.-]+\.[a-z0-9]+?)([/:]|$)', address).group(2)
+            if not self.in_whitelist(dom):
+                if get_tld(url_normalize(dom), fail_silently=True):
+                    return url_normalize(dom)
+                else:
+                    return None
+        except:
+            return None
+
+    def in_whitelist(self, domain: str) -> bool:
+        for dom_clean in self.domain_whitelist:
+            if re.match(dom_clean, domain):
+                return True
+        return False
+
+    def get_data_from_text(self, text) -> list:
+        data = []
+        for sub in self.substitutions:  # allows for custom matches
+            text = text.replace(sub[0], sub[1])
+        tweet_text = text.split()
+        tweet_text = [x.strip().lower() for x in tweet_text]
+        for part in tweet_text:
+            domain = self.get_domain(part)
+            if domain:
+                data.append(url_normalize(domain))
+        return data
+
+    def get_classification_type(self, text):
+        # TO IMPLEMENT
+        return 'unknown'
+
+    def process(self):
+        report = self.receive_message()
+        data = []
+        text = utils.base64_decode(report["raw"])
+        classification_type = self.get_classification_type(text)
+        data.extend(self.get_data_from_text(text))
+        for url in data:
+            event = self.new_event(report)
+            event.add('raw', url, overwrite=True)
```

Comment:
* there shouldn't be the need to overwrite
* raw means raw and unparsed. here parsed and unparsed is the same.

Reply:
So should the raw be the thing that was sent from the collector?

Reply:
Or the part of the previous raw that has been parsed. E.g. in csv's it's the header and the line.